### PR TITLE
SK-868: Handle all DidDowngrade scenarios

### PR DESCRIPTION
**Pull request by lukeg90**
SK-868

### CODE REVIEW

File: src/views/DidDowngrade/DidDowngrade.tsx [hunk 1]
```diff
@@ -3,14 +3,12 @@ import { Link } from 'react-router-dom';
 import BN from 'bn.js';
 import { browser } from 'webextension-polyfill-ts';
 import { IDidDetails } from '@kiltprotocol/types';
-
-import { DidChain } from '@kiltprotocol/did';
+import { IChainDeposit } from '@kiltprotocol/did/lib/cjs/Did.chain';
```
Comment (arty-name) [inline src/views/DidDowngrade/DidDowngrade.tsx:6]: Importing from paths like this one is not good, like accessing some private property. I’ll show a better alternative below.

File: src/views/DidDowngrade/DidDowngrade.tsx [hunk 2]
```diff
@@ -50,7 +51,8 @@ function useCosts(
 } {
   // useSwrDataOrThrow doesn’t fit here because some calls are conditional
   const [fee, setFee] = useState<BN | undefined>();
-  const [deposit, setDeposit] = useState<BN | undefined>();
+  const [depositDid, setDepositDid] = useState<IChainDeposit>();
```
Comment (arty-name) [inline src/views/DidDowngrade/DidDowngrade.tsx:54]:
```suggestion
  const [depositDid, setDepositDid] = useState<Awaited<ReturnType<typeof getDepositDid>>>();
```

File: src/views/DidDowngrade/DidDowngrade.tsx [hunk 3]
```diff
@@ -61,13 +63,29 @@ function useCosts(
       }
 
       setFee(await getFee(did));
-      setDeposit(await getDeposit());
+      setDepositDid(await getDepositDid(did));
+      setDepositWeb3Name(await getDepositWeb3Name(did));
     })();
   }, [did]);
 
+  const deposit = useMemo(() => {
```
Comment (arty-name) [inline src/views/DidDowngrade/DidDowngrade.tsx:71]: I assume it didn’t work out then to improve this memo?

**Review by arty-name — APPROVED** [review]

File: src/views/DidDowngrade/DidDowngrade.tsx (hunk 3 above)
Comment (lukeg90) [reply src/views/DidDowngrade/DidDowngrade.tsx:71]: I forgot what your initial suggestion was, maybe you can remind me? I couldn't come up with a good solution.

Comment (arty-name) [reply src/views/DidDowngrade/DidDowngrade.tsx:71]:
```js 
const didDeposit = ... ? 0 : ...;
const nameDeposit = ... ? 0 : ...;
return didDeposit.add(nameDeposit);

**Review by rflechtner — APPROVED** [review]
LGTM, just some ideas

File: src/utilities/didDowngrade/didDowngrade.ts [hunk 4]
```diff
@@ -19,30 +19,55 @@ interface DidTransaction {
...
   );
 
-  const blockchain = await BlockchainApiConnection.getConnectionOrConnect();
-  const tx = await blockchain.signTx(keypair, didAuthorizedExtrinsic);
+  const web3name = await Web3Names.queryWeb3NameForDid(fullDidDetails.did);
+
+  let extrinsic: SubmittableExtrinsic;
+
+  if (web3name) {
+    const w3nRemoveExtrinsic = await Web3Names.getReleaseByOwnerTx();
```
[the remark this answers is not in the document]
Comment (lukeg90) [reply src/utilities/didDowngrade/didDowngrade.ts:41]: I first tried DID authorizing the batch extrinsic, but got this error:

`Error: The details for did did:kilt:4pehddkhEanexVTTzWAtrrfo2R7xPnePpuiJLC7shQU894aY do not contain the required keys for this operation`

[the remark this answers is not in the document]
Comment (lukeg90) [reply src/utilities/didDowngrade/didDowngrade.ts:41]: @ntn-x2 said that by design DIDs do not support signing arbitrary batches.

**Pull request merged**